packages/frontend/src/redux/reducers/sign/index.js

```diff
@@ -25,43 +37,55 @@ const sign = handleActions({
                 .reduce((totalAmount, amount) => totalAmount.add(new BN(amount)), new BN(0)).toString(),
             fees: {
                 transactionFees: '', // TODO: Calculate total fees
-                gasLimit: allActions
-                    .filter(a => Object.keys(a)[0] === 'functionCall')
-                    .map(a => a.functionCall.gas)
-                    .reduce((totalGas, gas) => totalGas.add(gas), new BN(0)).toString(),
+                gasLimit: calculateGasLimit(transactions),
                 gasPrice: '' // TODO: Where to get gas price?
             },
             sensitiveActionsCounter: allActions
                 .filter(a => ['deployContract', 'stake', 'deleteAccount'].indexOf(Object.keys(a)[0]) > -1)
                 .length
         };
     },
-    [signAndSendTransactions]: (state, { error, payload, ready }) => {
-
-        if (!ready) {
-            return {
-                ...state,
-                status: 'needs-confirmation'
-            };
-        }
-
-        if (error) {
-            return {
-                ...state,
-                status: 'error',
-                error: payload
-            };
-        }
-
+    [handleSignTransaction.pending]: (state, { error, payload, ready }) => ({
+        ...state,
+        status: 'in-progress'
+    }),
+    [handleSignTransaction.fulfilled]: (state, { error, payload, ready }) => ({
+        ...state,
+        status: 'success'
+    }),
+    [handleSignTransaction.rejected]: (state, { error, payload, ready }) => ({
+        ...state,
+        status: error.message.includes('Exceeded the prepaid gas')
+            ? 'retry-tx'
+            : 'error',
+        error: error
+    }),
+    [setSignTransactionStatus]: (state, { payload }) => {
         return {
             ...state,
-            status: 'success'
+            status: payload.status
         };
     },
-    [setSignTransactionStatus]: (state, { payload }) => {
+    [multiplyGas]: (state, { payload: { transactions: transactionsString } }) => {
+        const transactions = deserializeTransactionFromString(transactionsString);
```

Comment:
To re-state from a high level, this case is:
1. The user tried to submit a tx with the original amount of gas, but it failed due to insufficient gas attached.
2. The user clicked on our brand new button that allows them to resubmit the tx, but with 2x as much gas attached as the initial attempt had.
3. The second attempt to submit the tx failed also _even though we attached 2x as much gas as the first attempt_.

**Currently**, the user will again see the button allowing them to 'resubmit this with more gas attached'.
If they click that button, we will re-submit the TX a _third_ time, but this time with 2x as much gas as the _second attempt_  -- now they are attaching **4x** as much gas as the original allotment.

Consider that If we allow the user to make a _third_ attempt, it will be with **8x** the amount of gas attached as they did for their first attempt, and a fourth attempt will have **16x** as much gas attached as their first attempt did, and so on.

@stefanopepe could you weigh in on this behaviour? -- from the product perspective:
1. Should it be possible to repeat the re-submit attempt multiple times with ever increasing amounts of gas attached each time?
2. If yes for # 1 above, then should we limit how many times they can use the 'retry with 2x as much gas as the previous attempt' button, or should we leave this as an un-capped operation that would allow them to retry many times, each time doubling the gas from the previous attempt?  If they retry 8 times, they would be submitting with **512x** as much gas attached as their original attempt had.

Reply:
@MaximusHaximus @marcinbodnar I would assume that yes, we should iterate the operation of attaching more Tgas to the transaction - being careful to not to **exceed** and pay too much gas. So I would keep the increases in smaller steps, to not incur in invalid errors for excessive Tgas attached to transactions.

Examples:
- [this tx](https://explorer.near.org/transactions/2w52XqgrBVM1By2mKNRDuwUyfTXyUNfdPeq71nk9HY9Q) was paying 20Tgas, and failed
- [this one](https://explorer.near.org/transactions/BVfcDhwVbuF49oNmCqjxjGkQEbGCKiboME2dn5TvdR1C) paid 200Tgas and worked well
- [this other one](https://explorer.near.org/transactions/5tBWCKUr5c4iMNWTnLHhw3pvLjRUx5pQoiyymHUWFYdK) paid 300Tgas and worked great as well

This one below didn't work because of excessive gas attached to it (400Tgas) and threw an invalid transaction error even before sending it to the network:
```
pepe$ near call zkv_staketosupportprivacy.poolv1.near ping '{}' --accountId pinger.near --gas 400000000000000
Scheduling a call: zkv_staketosupportprivacy.poolv1.near.ping({})
An error occured
Error: {"kind":{"ActionsValidation":{"TotalPrepaidGasExceeded":{"limit":300000000000000,"total_prepaid_gas":400000000000000}}}}
    at Object.parseRpcError (/usr/local/lib/node_modules/near-cli/node_modules/near-api-js/lib/utils/rpc_errors.js:24:19)
    at /usr/local/lib/node_modules/near-cli/node_modules/near-api-js/lib/providers/json-rpc-provider.js:312:44
    at processTicksAndRejections (node:internal/process/task_queues:94:5)
    at async Object.exponentialBackoff [as default] (/usr/local/lib/node_modules/near-cli/node_modules/near-api-js/lib/utils/exponential-backoff.js:7:24)
    at async JsonRpcProvider.sendJsonRpc (/usr/local/lib/node_modules/near-cli/node_modules/near-api-js/lib/providers/json-rpc-provider.js:297:26)
    at async /usr/local/lib/node_modules/near-cli/node_modules/near-api-js/lib/account.js:114:24
    at async Object.exponentialBackoff [as default] (/usr/local/lib/node_modules/near-cli/node_modules/near-api-js/lib/utils/exponential-backoff.js:7:24)
    at async Account.signAndSendTransactionV2 (/usr/local/lib/node_modules/near-cli/node_modules/near-api-js/lib/account.js:110:24)
    at async scheduleFunctionCall (/usr/local/lib/node_modules/near-cli/commands/call.js:54:34)
    at async Object.handler (/usr/local/lib/node_modules/near-cli/utils/exit-on-error.js:52:9)
ServerError: {"kind":{"ActionsValidation":{"TotalPrepaidGasExceeded":{"limit":300000000000000,"total_prepaid_gas":400000000000000}}}}
    at Object.parseRpcError (/usr/local/lib/node_modules/near-cli/node_modules/near-api-js/lib/utils/rpc_errors.js:24:19)
    at /usr/local/lib/node_modules/near-cli/node_modules/near-api-js/lib/providers/json-rpc-provider.js:312:44
    at processTicksAndRejections (node:internal/process/task_queues:94:5)
    at async Object.exponentialBackoff [as default] (/usr/local/lib/node_modules/near-cli/node_modules/near-api-js/lib/utils/exponential-backoff.js:7:24)
    at async JsonRpcProvider.sendJsonRpc (/usr/local/lib/node_modules/near-cli/node_modules/near-api-js/lib/providers/json-rpc-provider.js:297:26)
    at async /usr/local/lib/node_modules/near-cli/node_modules/near-api-js/lib/account.js:114:24
    at async Object.exponentialBackoff [as default] (/usr/local/lib/node_modules/near-cli/node_modules/near-api-js/lib/utils/exponential-backoff.js:7:24)
    at async Account.signAndSendTransactionV2 (/usr/local/lib/node_modules/near-cli/node_modules/near-api-js/lib/account.js:110:24)
    at async scheduleFunctionCall (/usr/local/lib/node_modules/near-cli/commands/call.js:54:34)
    at async Object.handler (/usr/local/lib/node_modules/near-cli/utils/exit-on-error.js:52:9) {
  type: 'InvalidTxError',
  context: ErrorContext {
    transactionHash: '7PHyLUf97c6R3E1UoH7AJZqhjbdsW4bN8FPJwRTkA6yc'
  },
  kind: {
    ActionsValidation: {
      TotalPrepaidGasExceeded: { limit: 300000000000000, total_prepaid_gas: 400000000000000 }
    }
  }
}
We attempted to send transaction 7PHyLUf97c6R3E1UoH7AJZqhjbdsW4bN8FPJwRTkA6yc to NEAR, but something went wrong.
To see the transaction in the transaction explorer, please open this url in your browser
https://explorer.mainnet.near.org/transactions/7PHyLUf97c6R3E1UoH7AJZqhjbdsW4bN8FPJwRTkA6yc
Note: if the transaction was invalid (e.g. not enough balance), it will show as "Not started" or "Finalizing"
```

The logic in my opinion should:
1. try again the transaction adding 25Tgas units to the original Tgas and keep going up until 300Tgas, after which it should fail completely. 
2. If the transaction is _already coming_ with 300Tgas, go in the opposite direction, decreasing 25Tgas units until 150Tgas, after which fail completely

Reply:
As a reference, see my comment here: https://github.com/near/near-wallet/issues/2111#issuecomment-956479345